### PR TITLE
Cron-based autoscaling for EHPA

### PR DESCRIPTION
**Pull request by kitianFresh**

### CODE REVIEW

File: deploy/metric-adapter/apiservice.yaml [hunk 1]
```diff
@@ -24,4 +24,19 @@ spec:
...
+  name: v1beta1.external.metrics.k8s.io
+spec:
+  service:
+    name: metric-adapter
+    namespace: crane-system
+  group: external.metrics.k8s.io
+  version: v1beta1
+  insecureSkipTLSVerify: true
+  groupPriorityMinimum: 100
+  versionPriority: 100
```
Comment (qmhu) [inline deploy/metric-adapter/apiservice.yaml:42]: add ApiService for v1beta2

File: pkg/controller/ehpa/effective_hpa_controller.go [hunk 2]
```diff
@@ -130,8 +130,12 @@ func (c *EffectiveHPAController) UpdateStatus(ctx context.Context, ehpa *autosca
 	if !equality.Semantic.DeepEqual(&ehpa.Status, newStatus) {
 		klog.V(4).Infof("EffectiveHorizontalPodAutoscaler status should be updated, currentStatus %v newStatus %v", &ehpa.Status, newStatus)
 
+		// use patch to avoid conflict errors
```
Comment (qmhu) [inline pkg/controller/ehpa/effective_hpa_controller.go:133]: For code consistence, maybe just use `status.Update` is fine to me.

File: pkg/controller/ehpa/hpa.go [hunk 3]
```diff
@@ -243,6 +245,11 @@ func (c *EffectiveHPAController) GetHPAMetrics(ctx context.Context, ehpa *autosc
 		metrics = append(metrics, customMetricsForPrediction...)
 	}
 
+	// Construct cron external metrics for cron scale
+	if IsCronScaleEnabled(ehpa) {
```
Comment (qmhu) [inline pkg/controller/ehpa/hpa.go:249]: Better to move `GetCronMetricSpecsForHPA` to hpa.go

File: pkg/controller/ehpa/predict.go [hunk 4]
```diff
@@ -160,6 +160,10 @@ func IsPredictionEnabled(ehpa *autoscalingapi.EffectiveHorizontalPodAutoscaler)
 	return ehpa.Spec.Prediction != nil && ehpa.Spec.Prediction.PredictionWindowSeconds != nil && ehpa.Spec.Prediction.PredictionAlgorithm != nil
 }
 
+func IsCronScaleEnabled(ehpa *autoscalingapi.EffectiveHorizontalPodAutoscaler) bool {
```
Comment (qmhu) [inline pkg/controller/ehpa/predict.go:163]: s/IsCronScaleEnabled/CronEnabled/g

File: pkg/metricprovider/cron_metric_provider.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	ehpaInformer         cache.SharedIndexInformer
+
+	scalersLock sync.RWMutex
+	cronScalers map[string]*CronScaler
+}
+
+func NewCronMetricProvider(client client.Client, craneClient craneclientset.Interface) *CronMetricProvider {
+
+	craneInformerFactory := craneinformers.NewSharedInformerFactory(craneClient, informerSyncPeriod)
+	ehpaInformer := craneInformerFactory.Autoscaling().V1alpha1().EffectiveHorizontalPodAutoscalers().Informer()
```
Comment (qmhu) [inline pkg/metricprovider/cron_metric_provider.go:44]: s/ehpaInformer/autoscalingInformer/g

File: pkg/metricprovider/cron_metric_provider.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+	return c
+}
+
+func (p *CronMetricProvider) Start(stopCh <-chan struct{}) bool {
+	go p.craneInformerFactory.Start(stopCh)
+	return cache.WaitForNamedCacheSync("cron-metric-provider", stopCh, p.ehpaInformer.HasSynced)
+}
+
+func (p *CronMetricProvider) addEHPA(obj interface{}) {
+	ehpa, ok := obj.(*autoscalingapi.EffectiveHorizontalPodAutoscaler)
```
Comment (qmhu) [inline pkg/metricprovider/cron_metric_provider.go:67]: Your can referring to this controller pattern:
https://github.com/kubernetes/sample-controller/blob/master/controller.go

File: pkg/controller/ehpa/predict.go (hunk 4 above)
Comment (kitianFresh) [reply pkg/controller/ehpa/predict.go:163]: > s/IsCronScaleEnabled/CronEnabled/g

I keep it same style with the IsPredictionEnabled, so the IsPredictionEnabled need to rename to PredictionEnabled?

File: pkg/metricprovider/cron_metric_provider.go (hunk 6 above)
Comment (kitianFresh) [reply pkg/metricprovider/cron_metric_provider.go:67]: > Your can referring to this controller pattern: https://github.com/kubernetes/sample-controller/blob/master/controller.go

ok

File: deploy/metric-adapter/apiservice.yaml (hunk 1 above)
Comment (kitianFresh) [reply deploy/metric-adapter/apiservice.yaml:42]: > add ApiService for v1beta2

There is no v1beta2 version for External metrics

File: pkg/metricprovider/cron_metric_provider.go (hunk 5 above)
Comment (kitianFresh) [reply pkg/metricprovider/cron_metric_provider.go:44]: why ? this informer is just used for ehpa

Comment (qmhu) [reply pkg/metricprovider/cron_metric_provider.go:44]: ehpa and evpa are all included in this package.

File: pkg/controller/ehpa/predict.go (hunk 4 above)
Comment (qmhu) [reply pkg/controller/ehpa/predict.go:163]: ok, how about IsCronEnabled.

File: cmd/metric-adapter/main.go [hunk 7]
```diff
@@ -50,19 +49,13 @@ type MetricAdapter struct {
 	Message string
 }
 
-func (a *MetricAdapter) makeProvider(remoteAdapter *metricprovider.RemoteAdapter, config *rest.Config, client client.Client) (provider.CustomMetricsProvider, error) {
-	kubeClient, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return nil, fmt.Errorf("unable to create kube client: %v", err)
-	}
+func (a *MetricAdapter) makeProvider(remoteAdapter *metricprovider.RemoteAdapter, client client.Client, recorder record.EventRecorder) provider.CustomMetricsProvider {
```
Comment (mfanjie) [inline cmd/metric-adapter/main.go:52]: is initProvider better than makeProvider?
same comment to makeExternalMetricProvider

File: cmd/metric-adapter/main.go [hunk 8]
```diff
@@ -89,6 +83,7 @@ func main() {
 	cmd.Flags().IntVar(&remoteAdapterServicePort, "remote-adapter-service-port", 6443, "Port of remote adapter's service")
 	cmd.Flags().IntVar(&apiQps, "api-qps", 300, "QPS of rest config.")
 	cmd.Flags().IntVar(&apiBurst, "api-burst", 400, "Burst of rest config.")
+	cmd.Flags().IntVar(&controllerWorkers, "controller-workers", 4, "worker num of controller")
```
Comment (mfanjie) [inline cmd/metric-adapter/main.go:86]: controller-workers -> concurrent-workers?

File: cmd/metric-adapter/main.go (hunk 7 above)
Comment (kitianFresh) [reply cmd/metric-adapter/main.go:52]: > is initProvider better than makeProvider? same comment to makeExternalMetricProvider

why? What is the difference?

Comment (mfanjie) [reply cmd/metric-adapter/main.go:52]: just a minor comment, ok for either way, initProvider is what is used in kubernetes cloud provider

File: examples/autoscaling/effective-hpa-cron-local.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
+apiVersion: autoscaling.crane.io/v1alpha1
+kind: EffectiveHorizontalPodAutoscaler
+metadata:
+  name: php-apache-local
+spec:
```
Comment (qmhu) [inline examples/autoscaling/effective-hpa-cron-local.yaml:5]: How to ensure the timezone in metric-adapter container is sync to cluster and node? should we mount timezone config to metric-adapter ?

File: pkg/metricprovider/cron_metric_provider_test.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	return &str
+}
+
+func TestGetCronScaleLocation(t *testing.T) {
+	americaLoc, _ := time.LoadLocation("America/Adak")
+	asiaShanghaiLoc, _ := time.LoadLocation("Asia/Shanghai")
+	testCases := []struct {
+		desc         string
+		cronScale    autoscalingapi.CronSpec
+		wantLocation *time.Location
```
Comment (qmhu) [inline pkg/metricprovider/cron_metric_provider_test.go:21]: s/wantLocation/want/g

File: pkg/metricprovider/cron_metric_provider.go [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) handleUpdateObj(oldObj, newObj interface{}) {
+	item := &workItem{obj: newObj, oldObj: oldObj, itemType: workTypeUpdated}
+	klog.V(6).Infof("item %v updated", item)
+	p.workqueue.Add(item)
+}
+
+func (p *CronMetricProvider) reconcileEHPAWorkItem(key interface{}) error {
```
Comment (qmhu) [inline pkg/metricprovider/cron_metric_provider.go:232]: We don't recommend using workType to check which event type is. This is not guaranteed when event queue do full sync. Actual how to handle reconcile in controller is very familiar when we use controller-runtime and we don't check eventType is Add or Update or Delete in that.
Please refering to https://github.com/kubernetes/sample-controller/blob/master/controller.go#L243 to implement sync loop.

File: pkg/metricprovider/cron_metric_provider.go [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+	klog.V(6).Infof("EHPA %v starting updated", klog.KObj(newEhpa))
+	func() {
+		p.scalersLock.Lock()
+		defer p.scalersLock.Unlock()
+
+		for _, name := range oldNames {
+			delete(p.cronScalers, name)
+		}
+		for _, scaler := range newScalers {
+			p.cronScalers[scaler.Name()] = scaler
```
Comment (qmhu) [inline pkg/metricprovider/cron_metric_provider.go:322]: Do we need to save cronScalers in Provider?  
What if we just check cron status for each request and not save it in Provider.

File: examples/autoscaling/effective-hpa-cron-local.yaml (hunk 9 above)
Comment (kitianFresh) [reply examples/autoscaling/effective-hpa-cron-local.yaml:5]: There are three ways to setting the container timezone.  it is depends on how to deploy the service. 
1. setting the timezone when build the docker image.
2. use host path to pass the host's timezone to container.
3. use env to set the pod yaml to set the timezone.

currently i use build image default timezone. the first one.  I think use 3 is better.

File: pkg/metricprovider/cron_metric_provider.go (hunk 12 above)
Comment (kitianFresh) [reply pkg/metricprovider/cron_metric_provider.go]: Yes, i removed the cache

**Pull request merged**